okhttp/src/main/java/okhttp3/internal/publicsuffix/PublicSuffixDatabase.java

```
@@ -0,0 +1,325 @@
+  // that will likely never be used. Each rule is separated by '\n'. Please see the
+  // PublicSuffixListGenerator class for how these lists are generated.
+  // Guarded by this.
+  private byte[] publicSuffixListBytes;
+  private byte[] publicSuffixExceptionListBytes;
+
+  /**
+   * Returns the effective top-level domain plus one (eTLD+1) by referencing the public suffix list.
+   * Returns null if the domain is a public suffix or the list was unable to be loaded.
+   *
+   * <p>Here are some examples: <pre>{@code
+   * assertEquals("google.com", getEffectiveTldPlusOne("google.com"));
+   * assertEquals("google.com", getEffectiveTldPlusOne("www.google.com"));
+   * assertNull(getEffectiveTldPlusOne("com"));
+   * }</pre>
+   */
+  public String getEffectiveTldPlusOne(String domain) {
+    if (domain == null) return null;
+
+    // Confirm we're dealing with a valid domain.
+    // TODO: Domains that start with '.' will be null on JDK8, but not on JDK7.
+    String canonicalDomain = Util.domainToAscii(domain);
+    if (canonicalDomain == null || canonicalDomain.startsWith(".")) return null;
+
+    // Convert it back to Unicode because the list uses Unicode, not punycode form.
+    String unicodeDomain = IDN.toUnicode(canonicalDomain);
+
+    // Preserve the passed in domain format (Unicode vs. punycode), but canonicalize it. This
+    // satisfies the publicsuffix.org test cases, but it's unclear if this should be done generally.
+    String domainToReturn = canonicalDomain.equalsIgnoreCase(domain)
+        ? canonicalDomain // We were passed a punycode domain.
+        : unicodeDomain; // We were passed a Unicode domain.
+
+    String[] domainLabels = unicodeDomain.split("\\.");
+    String[] rule = findMatchingRule(domainLabels);
+
+    if (rule == null
+        || (domainLabels.length == rule.length && rule[0].charAt(0) != EXCEPTION_MARKER)) {
+      // We failed to read the list or the domain is a public suffix.
```

K, good point. I'd wanna find that out in a test environment and failing silently doesn't do that. Will fix.